### ISSUE TITLE
Get error messages indexed by name

### ISSUE THREAD

**Issue by batusa** [Respect/Validation#664]
taking a similar example from #649 :

``` php
try {
    v::key('activity', v::notEmpty()->alnum()->length(10))
      ->key('department', v::notEmpty())
      ->assert(['activity' => '', 'department' => '']);
} catch(NestedValidationException $e) {
    print_r($e->getMessagesIndexedByName()); // or just getMessagesByName() ?
}
```

the code above would reproduce:

```
[activity] => Array
    (
        [0] => activity must not be empty
        [1] => activity must contain only letters (a-z) and digits (0-9)
        [2] => activity must have a length greater than 10
    ) 
[department] => Array
    (
        [0] => department must not be empty
    )
```

``` php
try {
    v::alnum()->length(6, 12)->setName('username')->assert('@');
} catch(NestedValidationException $e) {
    print_r($e->getMessagesIndexedByName());
}
```

the code above would reproduce:

```
[username] => Array
    (
        [0] => username must contain only letters (a-z) and digits (0-9)
        [1] => username must have a length between 6 and 12
    )
```

so far the only problem I've seen would be in case the input is empty/null and the user don't set a specific name with `setName('name')` then the arrays' indexes would be empty too.

@henriquemoody what do you think about it?

**Comment by aguilera51284** [comment]
i have the same problem :S

**Comment by JCMais** [comment]
This is what is getting me from using this awesome library. :disappointed: 

Showing error messages by value is not intuitive. Take a form with N required fields. If the user submit almost all of them empty, he is going to get many messages saying
```
'' must not be empty
'' must not be empty
'' must not be empty
'' must not be empty
```

Which is basically useless.

**Comment by aight8** [comment]
This issue makes the whole library useless sorry.

**Comment by schnittstabil** [comment]
I really like the idea, but how should nested arrays be handled? E.g.:

```php
try {
	v::key(
		'con.tact', v::key('num.ber', v::notEmpty())
	)->assert([
		'con.tact'=> ['num.ber' => '']
	]);
} catch(NestedValidationException $exception) {
	print_r($exception->getMessagesIndexedByName());
}
```
```php
[???] => Array
    (
        [0] => ??? num.ber must not be empty
    ) 
```

**Comment by rifton007** [comment]
Hi,
Any news about this feature ?

**Comment by codeaid** [comment]
So it's been open for 2 years and still no response? Installed the library thinking that this is it, I've found the one only to realize that error messages are absolutely useless. There's no way I can return a payload from an API explaining what error messages each field has, like e.g. Zend Validator.

Speaking about nested array values - simply nest them in their respective keys. E.g. with payload like this where `leaf` is expected to be an integer:

```json
{
    "root": {
        "branch": {
            "leaf": "notint"
        }
    }
}
```

...the error messages could look something like

```
[
    "root" => [
        "branch" => [
            "leaf" => [
                "notInt" => "Input is not an integer value",
                "notBetween" => "Input must be between 1 and 10",
            ],
        ],
    ],
];
```

Then it's up to the user to process those messages but at least they're structured in a destructible way. Also, setting custom message overrides is a must in my view. Currently without such feature, as @aight8 said, this library is pretty unusable.

**Comment by henriquemoody** [comment]
> Currently without such feature, as @aight8 said, this library is pretty unusable.

I have to disagree and if you check how much people are [using it in](https://packagist.org/packages/respect/validation/stats) your affirmation seems a little exaggerated. 

I'm working on it for quite some time in #678, and I got quite some juicy results. But since there are too many rules to refactor and I don't have as much time as I used to have, this is been postponed. So, if you want to speed up the process, help us out on #921.

**Comment by henriquemoody** [comment]
Fixed by #957.

I've changed how `getMessages()`, check out [get_messages.phpt](https://github.com/Respect/Validation/blob/master/tests/integration/get_messages.phpt) and [get_messages_with_replacements.phpt](https://github.com/Respect/Validation/blob/master/tests/integration/get_messages_with_replacements.phpt) to see how it works.

It will be released with the version 2.0. If any of you want to speed things up take a look at #921!

Thanks for reporting! :panda_face:
